packages/blade/src/components/Collapsible/CollapsiblePanel.web.tsx

```diff
@@ -0,0 +1,127 @@
+import type { ReactElement, ReactNode, TransitionEventHandler } from 'react';
+import { useRef } from 'react';
+import styled from 'styled-components';
+import { useCollapsibleContext } from './CollapsibleContext';
+import { castWebType, makeMotionTime, makeSize } from '~utils';
+import { useDidUpdate } from '~src/hooks/useDidUpdate';
+
+type CollapsiblePanelProps = {
+  children: ReactNode;
+};
+
+type StyledCollapsiblePanelProps = {
+  defaultIsExpanded: boolean;
+  isExpanded: boolean;
+};
+
+/**
+ * We can't animate to and from `auto` height, these are used for imperative css transitions.
+ * Overall on expanding height will change as: 0px -> Actual height -> auto
+ */
+const HEIGHT_EXPANDED = 'auto';
+const HEIGHT_COLLAPSED = '0px';
+
+// TODO: move common styles when implementing native
+const StyledCollapsiblePanel = styled.div<StyledCollapsiblePanelProps>((props) => {
+  const { theme, defaultIsExpanded, isExpanded } = props;
+  return {
+    transitionDuration: castWebType(makeMotionTime(theme.motion.duration.xmoderate)),
+    transitionTimingFunction: castWebType(theme.motion.easing.standard.effective),
+    transitionProperty: 'height, opacity',
+    opacity: isExpanded ? 1 : 0.8,
+    /**
+     * We need height explicitly here for initial styles because the component might be rendered on server,
+     * in which case for expanded items this should be `auto` because we don't know the actual pixel value.
+     *
+     * This is based on `defaultIsExpanded` rather than `isExpanded` because post initial render inline styles take over.
+     * Otherwise, changing `height` both here and in inline styles sometimes causes animation flickers due to styles mismatch.
+     */
+    height: defaultIsExpanded ? HEIGHT_EXPANDED : HEIGHT_COLLAPSED,
+    display: defaultIsExpanded ? 'block' : 'none',
+    overflowY: 'hidden',
+  };
+});
+
+const CollapsiblePanel = ({ children }: CollapsiblePanelProps): ReactElement => {
+  const { isExpanded, defaultIsExpanded } = useCollapsibleContext();
+  const collapsiblePanelRef = useRef<HTMLDivElement>(null);
+
+  /**
+   * This effect imperatively updates height to make css transitions work:
```

Comment:
i can see a lag when exapanding and collapsing animation. is that intentional and to be fixed in follow up PRs?




https://github.com/razorpay/blade/assets/11384858/70894827-09a1-4731-8f22-db8f54b37fd8

Reply:
Right, fixed that one in subsequent PRs. Please check the final interactions in the latest PR https://github.com/razorpay/blade/pull/1296